### PR TITLE
fix for the generics error caused due to the change in dart lang spec…

### LINKED ISSUE #184: Error with Latest Flutter Master Branch
```
Compiler message:
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:11:7: Error: Found unsupported uses of 'T' in supertype 'NotificationHandlers'.
class Interceptors<T>
      ^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:10:7: Error: Found unsupported uses of 'TNotification' in supertype 'NotificationHandlers'.
class Listeners<TNotification>
      ^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core/observable.dart:49:20: Error: The getter 'hasHandlers' isn't defined for the class 'Listeners<ChangeNotification<T>>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
 - 'ChangeNotification' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'hasHandlers'.
    if (_listeners.hasHandlers) {
                   ^^^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core/observable.dart:63:23: Error: The getter 'hasHandlers' isn't defined for the class 'Interceptors<T>'.
 - 'Interceptors' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'hasHandlers'.
    if (_interceptors.hasHandlers) {
                      ^^^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:14:48: Error: Too many positional arguments: 0 allowed, but 1 found.
Try removing the extra positional arguments.
  Interceptors(ReactiveContext context) : super(context);
                                               ^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:17:52: Error: Method not found: 'add'.
  Dispose intercept(Interceptor<T> interceptor) => add(interceptor);
                                                   ^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:17:52: Error: The method 'add' isn't defined for the class 'Interceptors<T>'.
 - 'Interceptors' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing method, or defining a method named 'add'.
  Dispose intercept(Interceptor<T> interceptor) => add(interceptor);
                                                   ^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:20:10: Error: Method not found: '_canHandle'.
    if (!_canHandle(change)) {
         ^^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:26:33: Error: Getter not found: '_handlers'.
      for (final interceptor in _handlers.toList(growable: false)) {
                                ^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:24:12: Error: Getter not found: '_context'.
    return _context.untracked(() {
           ^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:20:10: Error: The method '_canHandle' isn't defined for the class 'Interceptors<T>'.
 - 'Interceptors' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing method, or defining a method named '_canHandle'.
    if (!_canHandle(change)) {
         ^^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:24:12: Error: The getter '_context' isn't defined for the class 'Interceptors<T>'.
 - 'Interceptors' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named '_context'.
    return _context.untracked(() {
           ^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/interceptable.dart:26:33: Error: The getter '_handlers' isn't defined for the class 'Interceptors<T>'.
 - 'Interceptors' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named '_handlers'.
      for (final interceptor in _handlers.toList(growable: false)) {
                                ^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:12:45: Error: Too many positional arguments: 0 allowed, but 1 found.
Try removing the extra positional arguments.
  Listeners(ReactiveContext context) : super(context);
                                            ^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:14:65: Error: Method not found: 'add'.
  Dispose registerListener(Listener<TNotification> listener) => add(listener);
                                                                ^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:14:65: Error: The method 'add' isn't defined for the class 'Listeners<TNotification>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing method, or defining a method named 'add'.
  Dispose registerListener(Listener<TNotification> listener) => add(listener);
                                                                ^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:17:10: Error: Method not found: '_canHandle'.
    if (!_canHandle(change)) {
         ^^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:22:30: Error: Getter not found: '_handlers'.
      for (final listener in _handlers.toList(growable: false)) {
                             ^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:21:5: Error: Getter not found: '_context'.
    _context.untracked(() {
    ^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:17:10: Error: The method '_canHandle' isn't defined for the class 'Listeners<TNotification>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing method, or defining a method named '_canHandle'.
    if (!_canHandle(change)) {
         ^^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:21:5: Error: The getter '_context' isn't defined for the class 'Listeners<TNotification>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named '_context'.
    _context.untracked(() {
    ^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/listenable.dart:22:30: Error: The getter '_handlers' isn't defined for the class 'Listeners<TNotification>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named '_handlers'.
      for (final listener in _handlers.toList(growable: false)) {
                             ^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections/observable_list.dart:303:23: Error: The method 'add' isn't defined for the class 'Listeners<ListChange<T>>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
 - 'ListChange' is from 'package:mobx/src/api/observable_collections.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections.dart').
Try correcting the name to the name of an existing method, or defining a method named 'add'.
    return _listeners.add(listener);
                      ^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections/observable_map.dart:50:50: Error: The getter 'hasHandlers' isn't defined for the class 'Listeners<MapChange<K, V>>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
 - 'MapChange' is from 'package:mobx/src/api/observable_collections.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'hasHandlers'.
      _listenersField != null && _listenersField.hasHandlers;
                                                 ^^^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections/observable_map.dart:174:32: Error: The method 'add' isn't defined for the class 'Listeners<MapChange<K, V>>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
 - 'MapChange' is from 'package:mobx/src/api/observable_collections.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections.dart').
Try correcting the name to the name of an existing method, or defining a method named 'add'.
    final dispose = _listeners.add(listener);
                               ^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections/observable_set.dart:53:50: Error: The getter 'hasHandlers' isn't defined for the class 'Listeners<SetChange<T>>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
 - 'SetChange' is from 'package:mobx/src/api/observable_collections.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'hasHandlers'.
      _listenersField != null && _listenersField.hasHandlers;
                                                 ^^^^^^^^^^^
file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections/observable_set.dart:125:32: Error: The method 'add' isn't defined for the class 'Listeners<SetChange<T>>'.
 - 'Listeners' is from 'package:mobx/src/core.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/core.dart').
 - 'SetChange' is from 'package:mobx/src/api/observable_collections.dart' ('file:///C:/flutter/flutter/.pub-cache/hosted/pub.dartlang.org/mobx-0.2.0/lib/src/api/observable_collections.dart').
Try correcting the name to the name of an existing method, or defining a method named 'add'.
    final dispose = _listeners.add(listener);
```

### PR DESCRIPTION
**Pull request by pavanpodila**
Fixes #184 

spec: https://github.com/dart-lang/language/blob/master/accepted/future-releases/contravariant-superinterface-2018/feature-specification.md

### CODE REVIEW

File: mobx/lib/src/interceptable.dart [hunk 1]
```diff
@@ -8,8 +8,7 @@ abstract class Interceptable<T> {
   Dispose intercept(Interceptor<T> interceptor);
 }
 
-class Interceptors<T>
-    extends NotificationHandlers<WillChangeNotification<T>, Interceptor<T>>
+class Interceptors<T> extends NotificationHandlers<WillChangeNotification<T>>
```
Comment (rrousselGit) [inline mobx/lib/src/interceptable.dart]: It should probably override `add` to take a `Interceptor<T>`, to make things type safe

Comment (pavanpodila) [reply mobx/lib/src/interceptable.dart]: Yup, great observation 👍

Comment (pavanpodila) [reply mobx/lib/src/interceptable.dart]: Added the checks and tests for it. Please review @rrousselGit

Comment (rrousselGit) [reply mobx/lib/src/interceptable.dart]: I would expect using `covariant` keyword instead:

```dart
class NotificationHandlers<TNotification> {
  VoidCallback add(covariant Function handler) {}
}


class Foo extends NotificationHandlers<Object> {
  @override
  VoidCallback add(ValueChanged<Object> handler) => super.add(handler);
}
```

This makes the change safe at compile time

**Comment by rrousselGit** [comment]
Besides the `covariant` comment, LGTM

**Comment by pavanpodila** [comment]
Awesome...can you approve and merge ?

**Comment by rrousselGit** [comment]
You don't want to do the changes describe here? https://github.com/mobxjs/mobx.dart/pull/190#discussion_r291505797

File: mobx/lib/src/interceptable.dart (hunk 1 above)
Comment (pavanpodila) [reply mobx/lib/src/interceptable.dart]: Ah missed this ....not sure how, but this comment did not show up earlier. Will do this change.

**Pull request merged**